core/src/window_service.rs

```diff
@@ -77,22 +80,46 @@ where
     let now = Instant::now();
     inc_new_counter_debug!("streamer-recv_window-recv", packets.packets.len());
 
-    let mut shreds = vec![];
-    let mut discards = vec![];
-    for (i, packet) in packets.packets.iter_mut().enumerate() {
-        if let Ok(s) = bincode::deserialize(&packet.data) {
-            let shred: Shred = s;
-            if shred_filter(&shred, &packet.data) {
-                packet.meta.slot = shred.slot();
-                packet.meta.seed = shred.seed();
-                shreds.push(shred);
-            } else {
-                discards.push(i);
-            }
-        } else {
-            discards.push(i);
-        }
-    }
+    let (shreds, discards): (Vec<_>, Vec<_>) = thread_pool.install(|| {
+        packets
+            .packets
+            .par_iter_mut()
+            .enumerate()
+            .map(|(i, packet)| {
```

Comment:
Can't you use packets.packets.into_par_iter().filter here?

Reply:
because packets is an inner object on "packets" (ugh). 
I can't move the inner packets out of the outer container easily (i can use drain and then par_bridge) but i still have the fold + reduce, its just slightly less manual than before.

wdyt?

Reply:
hm what about `filter_map` into `(should_keep: bool, Option<Shred>)`, and then run `retain` at the end like this:

`let keep = [false, true, true, false, true];
let mut i = 0;
packets.packets.retain(|_| (keep[i], i += 1).0);`